README.md
# Chat Room Server

## About
This script implement the basic function of the chat room. 

**Note:** It is unsecure, unflexible version. Wait for update!

## Usage

* Server-Side:
```python
python chatroom_server.py [PORT]
```
Default Port is 5000.

* Client-Side:
```bash
nc IP PORT
```

## Future Feature
* Client's Application Instead of `nc`.
* Colored Text.
* Private Channel.
* File Transmit.
* Use Configer File to Manage Configurations of Server's Socket.

## Issue
* Does not catch the exception while launch the server.


Changelog.md
## [0.0.1] - 2019-11-22

### Added

- Initialize the project.
- Implement the simple function of chat room server.

chat_message.py

HEADER = '\033[95m'
OKBLUE = '\033[94m'
OKGREEN = '\033[92m'
WARNING = '\033[93m'
FAIL = '\033[91m'
ENDC = '\033[0m'
BOLD = '\033[1m'
UNDERLINE = '\033[4m'

BANNER = R"""
--------------------------------------------------------------
_________ .__                __________                       
\_   ___ \|  |__ _____ ______\______   \ ____   ____   _____  
/    \  \/|  |  \\__  \\_  __ \       _//  _ \ /  _ \ /     \ 
\     \___|   Y  \/ __ \|  | \/    |   (  <_> |  <_> )  Y Y  \
 \______  /___|  (____  /__|  |____|_  /\____/ \____/|__|_|  /
        \/     \/     \/             \/                    \/ 
--------------------------------------------------------------
""".encode('utf-8')

COMMAND_LIST = \
b"""----------------------------
!help list all command. 
!l    list all online user.
!q    leave the chat.
----------------------------
"""

MESSAGE_DASH = b'-'*28 + b'\n'

CLIENT_GREETING_MESSAGE = b"[*] Type Your Name: "
CLIENT_WELCOME_MESSAGE = b"[!] Welcome to Chatroom!\n"
CLIENT_LEAVE_MESSAGE = b"[!] Bye Bye!\n"
CLIENT_HELP_MESSAGE = b"[!] Type !help to see all command.\n"

CHAT_JOIN_MESSAGE = "\n [ {} has Joined the Chat! 🎉 ]\n"
CHAT_LEAVE_MESSAGE = "\n [ {} has leaved the Chat! 🍺 ]\n"

chatroom_server.py
#! /usr/bin/python3

import socket
import threading
import sys

import chat_message as msg
import chat_command as cmd

# Constant Variable
HOST = '0.0.0.0'
PORT = int(sys.argv[1]) if len(sys.argv) > 1 else 5000
PAR = 10
SIZE = 1024
SERVER = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
SERVER.bind((HOST, PORT))

# Static Variable
ClientList = {}
ClientAddress = {}

def LOG(message) -> None:
	print("[LOG]:", message)

def acceptClient() -> None:
	while True:
		client_socket, client_address = SERVER.accept()
		LOG(f"Connected From {client_address[0]}:{client_address[1]}")
		client_socket.send(msg.BANNER)
		client_socket.send(msg.CLIENT_GREETING_MESSAGE)
		ClientAddress[client_socket] = client_address
		threading.Thread(target=handleClient, args=(client_socket, client_address)).start()
	pass

def handleClient(client_socket, client_address) -> None:
	display_name = client_socket.recv(SIZE).decode('utf-8').strip()
	client_socket.send(msg.CLIENT_WELCOME_MESSAGE)
	client_socket.send(msg.CLIENT_HELP_MESSAGE)
	broadcastMessage(msg.CHAT_JOIN_MESSAGE.format(display_name).encode('utf-8'))
	ClientList[client_socket] = display_name
	name_tag = f"[{display_name}]: ".encode('utf-8')

	while True:
		try:
			message = client_socket.recv(SIZE)
			LOG(message)

			# FIXME
			# Use dict switch instead of if-else
			if message == b"!help\n":
				cmd.help(client_socket)
			elif message == b"!q\n":
				cmd.quit(client_socket, ClientList)
				broadcastMessage(message=msg.CHAT_LEAVE_MESSAGE.format(display_name).encode('utf-8'))
				LOG(f"{client_address[0]}:{client_address[1]} is Leaved")
				break
			elif message == b'!g\n':
				cmd.greet()
			elif message == b'!l\n':
				cmd.list(client_socket, ClientList)
			elif message == b'':
				LOG("empty")
				client_socket.close()
				del ClientList[client_socket]
				broadcastMessage(message=msg.CHAT_LEAVE_MESSAGE.format(display_name).encode('utf-8'))
				break
			elif not message.isspace():
				broadcastMessage(name_tag, message)

		except:
			client_socket.close()
			del ClientList[client_socket]
			broadcastMessage(message=msg.CHAT_LEAVE_MESSAGE.format(display_name).encode('utf-8'))
			break
	pass

def broadcastMessage(NameTag=b"", message=b"") -> None:
	for client in ClientList:
		client.send(NameTag + message)
	pass

def mainProcess() -> None:
	MAIN_THREAD = threading.Thread(target=acceptClient).start().join()
	pass

if __name__ == '__main__':
	SERVER.listen(PAR)
	LOG(f"Hosting Port: {PORT}")
	LOG("Waiting for Connection...\n")
	mainProcess()
	LOG("Server Shutdown...\n")
	SERVER.close()


chat_server.py
#! /usr/bin/python3

import socket
import threading
import argparse

# Argument Parse

# Constant Variable
HOST = '0.0.0.0' if True else None
PORT = 5000
PAR = 5
SIZE = 1024
SERVER = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
SERVER.bind((HOST, PORT))


BANNER = R"""
--------------------------------------------------------------
_________ .__                __________                       
\_   ___ \|  |__ _____ ______\______   \ ____   ____   _____  
/    \  \/|  |  \\__  \\_  __ \       _//  _ \ /  _ \ /     \ 
\     \___|   Y  \/ __ \|  | \/    |   (  <_> |  <_> )  Y Y  \
 \______  /___|  (____  /__|  |____|_  /\____/ \____/|__|_|  /
        \/     \/     \/             \/                    \/ 
--------------------------------------------------------------
""".encode('utf-8')
COMMAND_LIST = b"!help\tlist all command.     \
				!l   \tlist all online user. \
				!q   \tleave the chat."

CLIENT_GREETING_MESSAGE = b"[*] Type Your Name: "
CLIENT_WELCOME_MESSAGE = b"[!] Welcome to Chatroom!\n"
CLIENT_LEAVE_MESSAGE = b"[!] Bye Bye!\n"

CHAT_JOIN_MESSAGE = "\n [ <{}> has Joined the Chat! 🎉 ]\n"
CHAT_HELP_MESSAGE = "\n [Type !help to see all command.] \n"
CHAT_LEAVE_MESSAGE = "\n [ <{}> has leaved the Chat! 🍺 ]\n"

# Static Variable
ClientList = {}
ClientAddress = {}

def LOG(message) -> None:
	print("[LOG]:", message)

def acceptClient() -> None:
	while True:
		client_socket, client_address = SERVER.accept()
		LOG("Connected From {}:{}".format(client_address[0], client_address[1]))
		client_socket.send(BANNER)
		client_socket.send(CLIENT_GREETING_MESSAGE)
		ClientAddress[client_socket] = client_address
		threading.Thread(target=handleClient, args=(client_socket, client_address)).start()
	pass

def handleClient(client_socket, client_address) -> None:
	display_name = client_socket.recv(SIZE).decode('utf-8').strip()
	client_socket.send(CLIENT_WELCOME_MESSAGE)
	broadcastMessage(CHAT_JOIN_MESSAGE.format(display_name).encode('utf-8'))
	ClientList[client_socket] = display_name
	name_tag = "[{}]: ".format(display_name).encode('utf-8')

	while True:
		message = client_socket.recv(SIZE)
		LOG('-'*20)
		LOG(message)

		if message == b"!help\n":
			client_socket.send(COMMAND_LIST)
		elif message == b"!q\n":
			client_socket.send(CLIENT_LEAVE_MESSAGE)
			client_socket.close()
			del ClientList[client_socket]
			broadcastMessage(message=CHAT_LEAVE_MESSAGE.format(display_name).encode('utf-8'))
			LOG("{}:{} is Leaved".format(client_address[0], client_address[1]))
			break
		elif message == b'!l\n':
			# print(ClientList)
			client_socket.send('-'.encode('utf-8')*20 + b'\n')
			for k, v in ClientList.items():
				client_socket.send(b'- ' + v.encode('utf-8') + b'\n')
			client_socket.send('-'.encode('utf-8')*20 + b'\n')
		elif not message.isspace():
			broadcastMessage(name_tag, message)
	pass

def broadcastMessage(NameTag=b"", message=b"") -> None:
	for client in ClientList:
		client.send(NameTag + message)
	pass

def mainProcess() -> None:
	MAIN_THREAD = threading.Thread(target=acceptClient).start().join()
	pass

if __name__ == '__main__':
	SERVER.listen(PAR)
	LOG("Waiting for Connection...\n")
	mainProcess()
	LOG("Server Shutdown...\n")
	SERVER.close()
